[2단계 - 자동차 경주 리팩토링] 아마란스(송세연) 미션 제출합니다

amaran-th opened the pull request:
안녕하세요 준님, 아마란스입니다.
1단계 리뷰에 감사드립니다.
리뷰해주신 내용과 미션 가이드의 내용(MVC 구조)을 토대로 리팩토링을 진행하였습니다.

리팩토링 과정에서 한 가지 궁금한 점이 생겨 질문 드립니다.
1단계 미션에서 view 클래스의 InputView, OutputView의 메서드를 비 static으로 구현하였는데요, InputView와 OutputView와 같은 view 클래스의 경우 메서드를 전부 static으로 바꿔주는 게 좋을지, 아니면 지금 구현대로 가는 게 좋을지 고민이 됐습니다. 혹시 조언 주실 수 있으실까요?

Commit:
docs : README 파일에  2단계 미션 추가

- 리팩토링 요구사항 정리

(cherry picked from commit 17ceb223a97eef31e71653b4dfde3e47fa57ebb5)
(cherry picked from commit 433e87419e006f7795200789c4e5862c9f2bb2a0)

Commit:
refactor : Message 클래스 제거

- Message 클래스를 사용하는 것 대신 InputView, OutputView에서 직접 텍스트를 넣어주도록 수정

(cherry picked from commit 7524998831787e66d2392677ac787a30cf56078b)
(cherry picked from commit 333305510d5ad58b543f4c52ebaa9224d3ddaf40)

Commit:
refactor(CarTest) : 파일 끝 개행문자 추가

(cherry picked from commit 6dd0545156968f028be2ba5e9b277f5c31ae468e)
(cherry picked from commit 01060962c592c66a07e17949d819a71ebb3040ff)

Commit:
refactor(PowerSupllier) : RandomUtil을 strategy 패턴으로 구현

- PowerSupplier라는 인터페이스와 RandomPowerSupplier라는 구현체를 만들어 구현하였다.
- 테스트를 위한 LowPowerSupplier, HighPowerSupplier 구현체 생성

(cherry picked from commit bab9bd8489fa4020022a55748bf065fc1a299b23)
(cherry picked from commit c47aba447179a6207d4de9a53be464a097cdad28)

Commit:
refactor(RaceGame) : 전체적인 클래스 구조 수정

- RaceController에 main 함수 구현
- RaceGame의 race 메서드가 시도 횟수와 자동차 이름 목록을 매개변수로 받도록 구현

(cherry picked from commit b423d8977f41edb50680177b3ff35e8e9936ebad)
(cherry picked from commit 9b38f4107d95fd04db416e11b409016d7dd89d18)

Commit:
refactor(RaceGame) : 불필요한 sort 구문 제거

(cherry picked from commit 4fc8e5af3300248008a34ba3abbc9cdd430027b8)
(cherry picked from commit 870687db272a37e6ca51c0c6b277694c54b094b9)

Commit:
refactor : 형식에 맞춰 상수/메서드의 제어자 수정

(cherry picked from commit 87fcb719e0e72b40cd8b4cfa79fcf1c8b54c2ff8)
(cherry picked from commit 829011d44da03660bb0a15e1f4fa3bbb15b2c05b)

Commit:
refactor(Test) : 자동차 이동/멈춤을 strategy 구현체로 구현

(cherry picked from commit 399a0bf587ef05ba868feb99476e9a3b6be0272c)
(cherry picked from commit 16ac8b9fedcb60529b4b6fcb2adbd624cffb4ce7)

Commit:
refactor : 컨벤션 맞추기

- 파일 끝 개행문자 추가

(cherry picked from commit ed145cc10315e3985f95e8b14f0a932330543dd2)
(cherry picked from commit a9283ec296cd61d5f7c69fced2a6dc6e5db69aad)

choijy1705 commented on src/main/java/contoller/RaceController.java:
```diff
@@ -1,70 +1,31 @@
-  public void run() {
-    List<Car> cars = makeCars();
-    int roundNum = repeat(inputView::readTrialNum);
+  public static void main(String[] args) {
```
Application을 지우고 RaceController를 main으로 잡으셨네요
보통 Application 클래에서 main 을 호출하고 application에서 Controller를 생성하여 실행하는 방향이 일반적인 흐름일것 같아요
한 프로젝트에 컨트롤러가 여러개가 생기는 경우가 있을텐데 Controller안에 main이 숨어있다면
어디서 실행하는지 모르는 문제가 있겠네요!

choijy1705 commented on src/main/java/contoller/RaceController.java:
```diff
@@ -1,70 +1,31 @@
 
-  private InputView inputView = new InputView();
-  private OutputView outputView = new OutputView();
+  private static InputView inputView = new InputView();
```
main함수를 controller에서 실행하게 함으로써 어쩔수없이 static으로 변경한것으로 보여지는데 맞을까요??
View를 static 으로 쓰려면 view 안의 모든 메서드들을 static으로 하고 별도로 생성자로 생성하지 않고 호출하도록 하는 방향이 좋을것 같아요.

choijy1705 commented on src/main/java/contoller/RaceController.java:
```diff
@@ -1,70 +1,31 @@
   }
 
-  private <T> T repeat(Supplier<T> inputReader) {
+  private static <T> T repeat(Supplier<T> inputReader) {
```
contorller 에서 static main을 실행하다보니 
모든 메서드들이 static 으로 바꿔야하는 문제가 있겠네요

choijy1705 commented on src/main/java/domain/RaceGame.java:
```diff
@@ -0,0 +1,60 @@
+  private static final OutputView outputView = new OutputView();
+  private static final PowerSupplier powerSupplier = new RandomPowerSupplier();
+
+  private List<String> runnerNames;
```
runnerNames는 cars를 생성할때 말고는 안쓰이는것 같아서 
굳이 변수로 할당하지 않아도 될것 같아요

choijy1705 commented on src/main/java/domain/RaceGame.java:
```diff
@@ -0,0 +1,60 @@
+  }
+
+  public void race() {
+    outputView.printStart(cars);
```
RaceGame은 domain인데 여기서 view를 호출한다면
mvc패턴에 어긋나지 않을까요?

choijy1705 commented on src/main/java/strategy/RandomPowerSupplier.java:
```diff
@@ -0,0 +1,14 @@
+
+import java.util.Random;
+
+public class RandomPowerSupplier implements PowerSupplier {
```
PowerSupplier 는 뭘 의미하는걸까요?
랜덤숫자를 생성하는건데 의미가 적절한가요?

choijy1705 replied:
이전 randomUtil이 훨씬 코드 이해하기 좋고 명확했던것 같은데
요 방식으로 바꾼 이유가 있을까요??😭

amaran-th replied:
프로그램 요구사항이 '자동차'의 이동 여부를 결정짓는 숫자다 보니 '연료'같은 개념으로 생각해서 그렇게 이름 지어줬습니다! 또 Util은 **여러** 클래스에서 사용하는 기능을 모아두는 게 좋다는 의견을 들어서 패키지 구조를 수정해줬습니다...!
기존 코드로 다시 수정하는 편이 더 좋을까요??

choijy1705 replied:
여러 사람과 같이 작업한다면 메서드명, 변수명등 이름도 명확하게 해주면 좋을것 같아요
PowerSupplier라는 이름은 너무 추상적인것 같은데 어떻게 생각하시나요
자동차 경주 게임의 요구사항을 모르는 사람이 봤을때 쉽게 이해할 수 있을까요?

이전코드와 지금코드중 어떤게 이해하기 쉬울지 고민해보면 좋을것 같아요

amaran-th replied:
평소 학교 과제같은 데에서 사용됐던 네이밍이 요구사항에 나온 표현을 사용했던 경우가 많아서 그렇게 작명했었는데 지금보니 개발자 입장에선 추상적인 표현이 맞는 것 같습니다.
Strategy 패턴은 유지하고 싶어 구조는 그대로 가고 인터페이스/구현체 이름만 수정했습니다...!

choijy1705 commented on src/test/java/domain/CarTest.java:
```diff
@@ -15,23 +29,22 @@ class CarTest {
   @Test
   void 자동차_이동() {
-    car1.move(7);
+    car1.move(highPowerSupplier.supply());
```
테스트 코드는 문서의 역할도 합니다
다른 사람들이 내가 작성한 코드의 기능이 어떤걸 하는지 쉽게 파악할 수 있어서요
기존의 방식이 처음 보는 사람이 봐도 훨씬 이해하기 좋을것 같아요!

amaran-th replied:
숫자 값을 하드코딩하는 것보다 highPowerSupplier라는 표현이 더 이해하기 쉬울 거라 생각해서 그렇게 변경했습니다...!
highPowerSupplier.supply()를 따로 highPower라는 변수로 분리하면 괜찮을까요??

choijy1705 reviewed (changes requested):
안녕하세요 아마란스 
많은 시도를 해주셨네요!
몇 가지 리뷰 남겼는데 확인해주세요!

Commit:
refactor(controller) : 폴더 구조 및 파일명 수정

- 기존 RaceController 클래스의 이름을 Application으로 수정
- RaceGame을 도메인에서 controller 패키지로 이동

Commit:
refactor(view) : view의 메서드를 static 메서드로 변경

Commit:
fix(CarTest) : 테스트가 제대로 수행되지 않는 문제 해결

- 연료(값)을 좀 더 이해하기 쉽도록 변수명을 지어줌

Commit:
refactor(RaceGame) : runnerNames 멤버변수 제거

choijy1705 commented on src/main/java/Application.java:
```diff
@@ -1,9 +1,26 @@
+
+    final RaceGame raceGame = new RaceGame(carNames, roundCount);
+    raceGame.race();
+    OutputView.printWinners(raceGame.pickWinner());
```
controller 의 역할을 main함수에서 하고있는것 같아요!
MVC 에서 controller의 역할이 뭘까요?
Application에서는 RaceController 이외의 Controller가 있을 수 있어요
지금과 같은 형태라면 새로운 controller가 들어왔을때 추가하기 어려울것 같아요

choijy1705 commented on src/main/java/view/OutputView.java:
```diff
@@ -5,38 +5,38 @@
 
 public class OutputView {
```
View의 메서드들을 모두 static 메서드로 관리해주셨네요
이럴 경우
private 생성자를 만들어서 외부에서 OutputView 를 생성자를 통해 만들지 못하도록 해주면 좋을것 같아요
```java
  private OutputView() {
  }
```

choijy1705 commented on src/main/java/view/OutputView.java:
```diff
@@ -5,38 +5,38 @@
+  public static void printCarsStatus(List<Car> cars) {
+    cars.forEach(OutputView::printCarStatus);
     System.out.println();
   }
```
view 에서 domain에 대해서 알고 있다면
mvc 패턴의 의미가 모호해지는것 같아요
해당 forEach문은 controller에서 처리하면 cars를 파라미터로 안넘겨도 되지않을까요?

choijy1705 commented on src/main/java/view/OutputView.java:
```diff
@@ -5,38 +5,38 @@
-    System.out.printf(Messages.OUTPUT_CAR_STATUS, car.getName(), toBar(car.getPosition()));
+  private static void printCarStatus(Car car) {
+    System.out.printf("%s : %s\n", car.getName(), toBar(car.getPosition()));
   }
```
여기도 Car라는 도메인을 알고 있네요
view에서는 자동차이름과 position만 넘겨주면 되지않을까요?

choijy1705 reviewed (changes requested):
이전보다 조금더 이해하기 좋아진것 같아요👍
조금더 MVC 패턴에 맞게 리팩터링 해보면 좋을것 같아요
몇가지 더 코멘트 남겼으니 고민해보면 좋을것 같습니다

Commit:
refactor(view) : private 생성자 추가

Commit:
refactor(PowerSupplier) : 인터페이스/구현체 이름 변경

Commit:
refactor(OutputView) : mvc 모델 보완

- printCarsStatus 로직을 RaceGame으로 옮김

Commit:
refactor(RaceGame) : main함수의 controller 부분을 controller 코드로 이동

- RaceGame을 RaceController로 클래스명 변경

choijy1705 commented on src/test/java/domain/CarTest.java:
```diff
@@ -15,23 +29,24 @@ class CarTest {
   @Test
   void 자동차_이동() {
-    car1.move(7);
+    int highPower = highPowerSupplier.generate();
```
test를 위해서 클래스를 안만들고 그냥 원하는 숫자를 넣으면
충분히 테스트가 될것 같아요

choijy1705 commented on src/test/java/strategy/HighNumberGenerator.java:
```diff
@@ -0,0 +1,8 @@
+package strategy;
+
+public class HighNumberGenerator implements NumberGenerator {
```
위 리뷰를 반영하신다면 
없어도 될 클래스가 되겠네요

choijy1705 reviewed (approved):
1단계 진행하느라 고생많으셨어요!
좀 더 이해하기 쉬워진것 같아요
이번 미션은 여기서 머지하겠습니다
